src/Widgets/NotificationEntry.vala

```diff
@@ -231,6 +265,28 @@ public class Notifications.NotificationEntry : Gtk.ListBoxRow {
 
         show_all ();
 
+        button_release_event.connect (() => {
+            unowned var action_group = get_action_group (ACTION_GROUP_PREFIX);
+            if (action_group.has_action (notification.default_action)) {
+                notification.app_info.launch_action (Notification.DEFAULT_ACTION, new GLib.AppLaunchContext ());
+                action_group.activate_action (notification.default_action, null);
+                clear (false);
+            } else if (notification.actions.length == 0) {
+                try {
+                    notification.app_info.launch (null, null);
+                } catch (Error e) {
+                    critical ("Unable to launch app: %s", e.message);
+                }
+                clear ();
+            } else {
+                return Gdk.EVENT_STOP;
+            }
```

Comment:
a cleaner way to deal with this is to use the `Gtk.Actionable` interface of `Gtk.ListBoxRow`. and deal with the non-action case in the `activate()` signal.

Reply:
Hmm we would still need the button_release event handler to clear the notification, close the popover and launch the app info so there wouldn't be much gained there. Or is there a way I missed?

Reply:
like with the buttons case, i expect the NotificationMonitor to deal with the `ActionInvoked` signal that this will emit, and remove the notification there.

correct me if i'm wrong, but `Gtk.ListBox.row_activated()` is emitted even when using a action_name or activate() no? closing the popover can be handled there.

Reply:
> correct me if i'm wrong, but `Gtk.ListBox.row_activated()` is emitted even when using a action_name or activate() no? closing the popover can be handled there.

No it seems it doesn't but that's actually quite useful here so we can differentiate between activation via default action or without any action.